### PR TITLE
Issue #7269 - Refinement of `HostPort` behavior based on Specs

### LINKED ISSUE #7269: HostPort needs updates for spec compliance
**Jetty version(s)**
9.4+

**Java version/vendor** `(use: java -version)`
All

**OS type/version**
Alll

**Description**

Currently `HostPort` allows odd authorities that make no sense.

```
Host: :9999 
Host: -
Host: -:8888
Host: *
Host: *:2222
Host: *.eclipse.org
```

I think `HostPort` should validate the `host` portion a bit more, to reject these nonsense hosts with a 400 Bad Request.

The spec for HTTP at https://datatracker.ietf.org/doc/html/rfc7230#section-5.4
Says the `uri-host` for http is detailed in https://datatracker.ietf.org/doc/html/rfc7230#section-2.7.1
Which points to `host` in https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2

which has the following ABNF ...

```
  uri-host    = <host, see [RFC3986], Section 3.2.2>
  host        = IP-literal / IPv4address / reg-name
  IP-literal  = "[" ( IPv6address / IPvFuture  ) "]"
  IPv6address =                            6( h16 ":" ) ls32
                  /                       "::" 5( h16 ":" ) ls32
                  / [               h16 ] "::" 4( h16 ":" ) ls32
                  / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
                  / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
                  / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
                  / [ *4( h16 ":" ) h16 ] "::"              ls32
                  / [ *5( h16 ":" ) h16 ] "::"              h16
                  / [ *6( h16 ":" ) h16 ] "::"

  ls32        = ( h16 ":" h16 ) / IPv4address
                  ; least-significant 32 bits of address

  h16         = 1*4HEXDIG
                  ; 16 bits of address represented in hexadecimal
  IPvFuture   = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
  IPv4address = dec-octet "." dec-octet "." dec-octet "." dec-octet
  dec-octet   = DIGIT                 ; 0-9
                  / %x31-39 DIGIT         ; 10-99
                  / "1" 2DIGIT            ; 100-199
                  / "2" %x30-34 DIGIT     ; 200-249
                  / "25" %x30-35          ; 250-255
  reg-name    = *( unreserved / pct-encoded / sub-delims )
  unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
  pct-encoded = "%" HEXDIG HEXDIG
  sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
```

Optionally, this validation could exist as a `RejectInvalidAuthorityCustomizer` (like proposed in PR #7251).

### LINKED ISSUE #7285: HPackDecoderTest.testTooBigToIndex() has bad assumption on Host header value
**Jetty version(s)**
9.4.x+

**Java version/vendor** `(use: java -version)`
All

**OS type/version**
All

**Description**
See: https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-http2/http2-hpack/src/test/java/org/eclipse/jetty/http2/hpack/HpackDecoderTest.java#L251

https://github.com/eclipse/jetty.project/blob/b8b32d0fdd0286bb212b4deccb61474d72fe280e/jetty-http2/http2-hpack/src/test/java/org/eclipse/jetty/http2/hpack/HpackDecoderTest.java#L251

A Host on an authority cannot contain spaces like it's validating in the assumption.

### PR DESCRIPTION
**Pull request by joakime**
+ Updates to `HostPort` to satisfy spec rules on URI authority
+ Added `HttpURI.NO_PORT` in regards to handling of no-port to conform to `java.net.URI` behaviors of no-port as well (which uses -1, not 0)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java [hunk 1]
```diff
@@ -26,62 +26,122 @@
...
             Arguments.of("0:1:2:3:4:5:6", "[0:1:2:3:4:5:6]", null),
             Arguments.of("127.0.0.1:65535", "127.0.0.1", "65535"),
+            // Whitespace
+            Arguments.of("host   ", "host", null),
+            Arguments.of("    host   ", "host", null),
+            Arguments.of("host   :777", "host", 777),
+            Arguments.of("    host   :777", "host", 777),
+            Arguments.of("    host   :   777", "host", 777),
+            Arguments.of("    host   :777   ", "host", 777),
+            Arguments.of("    host   :   777   ", "host", 777),
```
Comment (joakime) [inline jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java:53]: These whitespace tests are odd, on one hand it's part of the "be lenient in what you accept", on the other hand, it's not to spec.

WDYT?

**Comment by joakime** [comment]
The failing test is `HPackDecoderTest.testTooBigIndex()`, which is now failing because that test has an assumption of an `:authority` (`Host`) header with spaces (which is invalid now in this PR)

I've opened an issue about this discovery at Issue #7285

**Comment by gregw** [comment]
@joakime I think you can fix the very large field test in this PR rather than make it a different issue/PR.  This PR tightens the requirements, so it should fix the tests it breaks.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 2]
```diff
@@ -184,9 +185,9 @@ String getMessage()
     public static HttpURI createHttpURI(String scheme, String host, int port, String path, String param, String query, String fragment)
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:185]: So what does it mean if port==0 is passed in?  Should that be corrected to NO_PORT? or left as 0?

File: jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java [hunk 3]
```diff
@@ -642,8 +644,8 @@ private boolean updateForwardedHandle(MethodHandles.Lookup lookup, String header
 
     private static class MutableHostPort
     {
-        public static final int UNSET = -1;
-        public static final int IMPLIED = 0;
+        public static final int UNSET = HostPort.NO_PORT;
+        public static final int IMPLIED = -2;
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java]:
```suggestion
        public static final int IMPLIED = HostPort.NO_PORT - 1;
```

File: jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java [hunk 4]
```diff
@@ -180,7 +180,7 @@ public void testLocalV2(RequestProcessor p) throws Exception
 
         assertThat(response, Matchers.containsString("HTTP/1.1 200"));
         assertThat(response, Matchers.containsString("pathInfo=/path"));
-        assertThat(response, Matchers.containsString("local=0.0.0.0:0"));
+        assertThat(response, Matchers.containsString("local=0.0.0.0:-1")); // LocalConnector has no local host or local port
```
Comment (gregw) [inline jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java]:
```suggestion
        assertThat(response, Matchers.containsString("local=0.0.0.0:" + HostPort.NO_PORT)); // LocalConnector has no local host or local port
```

File: jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java [hunk 5]
```diff
@@ -1002,7 +1002,7 @@ public boolean check(HttpServletRequest request, HttpServletResponse response)
         assertEquals("[::1]", results.get(i++));
         assertEquals("8888", results.get(i++));
         assertEquals("0.0.0.0", results.get(i++));
-        assertEquals("0", results.get(i));
+        assertEquals("-1", results.get(i)); // local connector has no local port
```
Comment (gregw) [inline jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java]:
```suggestion
        assertEquals(Integer.toString(HostPort.NO_PORT), results.get(i)); // local connector has no local port
```

File: jetty-util/src/main/java/org/eclipse/jetty/util/HostPort.java [hunk 6]
```diff
@@ -25,6 +25,7 @@
  */
 public class HostPort
 {
+    public static final int NO_PORT = -1;  // value used in java.net.URI for no-port
     private final String _host;
     private final int _port;
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/HostPort.java:30]: what if port==0 is passed in the constructor?

File: jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java (hunk 1 above)
Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java:53]: were we forgiving before?  If not, then lets not start being forgiving. If we were, then let's keep doing so

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java (hunk 2 above)
Comment (joakime) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:185]: port 0 is how we used to convey no port. (pre HttpURI object).
I'm tempted to reject port 0, but then again, there's cases where we create an HttpURI from something that isn't http. or on a network.
Either way, we shouldn't carry the port 0 thru to `java.net.URI` at all.

File: jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java:53]: we were not forgiving before.
I'll move these to invalid.

Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/HostPortTest.java:53]: Here's how to describe the old behavior...

* If the whitespace is in the _host_ portion of `HostPort`, it is preserved, and does not result in a 400 Bad Request, the request will reach the application, and use of various APIs will result in returning this whitespace, even secondary APIs will fail due to an invalid URL/URI during `getRequestURL()` or `sendRedirect()`.
* If the whitespace is in the _port_ portion of `HostPort`, the internal Integer.parseInt() fails with a `java.lang.NumberFormatException: For input string: "   777   "` and causes a 400 Bad Request.

**Comment by joakime** [comment]
@gregw I've pushed some of the updates from your recent review.

The only things I haven't touched is the whitespace changes ([see comment above](https://github.com/eclipse/jetty.project/pull/7279#discussion_r769146554) about how it behaves in 9.4.x before this PR) and how port=0 should be treated (or just documented).

**Comment by joakime** [comment]
@gregw would this one also be good for the new `jetty-12.0.x-hackathon` branch?

**Comment by joakime** [comment]
@gregw *bump* do you want to see this for `jetty-12.0.x` ?

**Comment by gregw** [comment]
There is definitely some goodness in this PR.  But I'd like a bit more clarity.

Can you update the PR description with exactly the changes that this PR is making?   From a quick look I can see:
 + handling of empty host by ????
 + validating characters used in host and if bad found then ????
 + improve handling of no port with the use of constants rather than 0 and/or -1
 + white space ignoring ???

We need a good list of the actual changes in behaviour.

**Comment by joakime** [comment]
No longer relevant for Jetty 9.4.x.

Moved discoveries from this PR to discussion at Issue #7269 for possible future addition to Jetty-12

**Pull request closed without merging**